lib/Dialect/FIRRTL/Transforms/HoistPassthrough.cpp

```diff
@@ -0,0 +1,571 @@
+  // Use local "hasDontTouch" to distinguish inner symbols on results
+  // vs on the operation itself (like an instance).
+  if (hasDontTouchOrInnerSymOnResult(v) ||
+      hasDontTouchOrInnerSymOnResult(ref.getValue()))
+    return {};
+  if (auto fop = ref.getValue().getDefiningOp<Forceable>();
+      fop && fop.isForceable())
+    return {};
+
+  // Limit to passive sources for now.
+  auto sourceType = type_dyn_cast<FIRRTLBaseType>(ref.getValue().getType());
+  if (!sourceType)
+    return {};
+  if (!sourceType.isPassive())
+    return {};
+
+  assert(hw::FieldIdImpl::getFinalTypeByFieldID(sourceType, ref.getFieldID()) ==
+             baseValue.getType() &&
+         "unexpected type mismatch, cast or extension?");
+
+  return HWDriver(connect, ref);
+}
+
+Value HWDriver::remat(PortMappingFn mapPortFn, ImplicitLocOpBuilder &builder) {
+  auto mappedSource = mapPortFn(getIndex(source.getValue()));
+  // TODO: Cast if needed.  For now only support matching.
+  // (No cast needed for current HWDriver's, getFieldRefFromValue and
+  // assert)
+  return getValueByFieldID(builder, mappedSource, source.getFieldID());
+}
+
+//===----------------------------------------------------------------------===//
+// MustDrivenBy analysis.
+//===----------------------------------------------------------------------===//
+namespace {
+/// Driver analysis, tracking values that "must be driven" by the specified
+/// source (+fieldID), along with final complete driving connect.
+class MustDrivenBy {
+public:
+  MustDrivenBy() = default;
+  MustDrivenBy(FModuleOp mod) { run(mod); }
+
+  /// Get direct driver, if computed, for the specified value.
+  Driver getDriverFor(Value v) const { return driverMap.lookup(v); }
+
+  /// Get combined driver for the specified value.
+  /// Walks the driver "graph" from the value to its ultimate source.
+  Driver getCombinedDriverFor(Value v) const {
+    Driver driver = driverMap.lookup(v);
+    if (!driver)
+      return driver;
+
+    // Chase and collapse.
+    Driver cur = driver;
+    size_t len = 1;
+    SmallPtrSet<Value, 8> seen;
+    while ((cur = driverMap.lookup(cur.source.getValue()))) {
+      // If re-encounter same value, bail.
+      if (!seen.insert(cur.source.getValue()).second)
+        return {};
+      driver.source = cur.source.getSubField(driver.source.getFieldID());
+      ++len;
+    }
+    (void)len;
+    LLVM_DEBUG(llvm::dbgs() << "Found driver for " << v << " (chain length = "
+                            << len << "): " << driver << "\n");
+    return driver;
+  }
+
+  /// Analyze the given module's ports and chase simple storage.
+  void run(FModuleOp mod) {
+    SmallVector<Value, 64> worklist(mod.getArguments());
+
+    DenseSet<Value> enqueued;
+    enqueued.insert(worklist.begin(), worklist.end());
+    while (!worklist.empty()) {
+      auto val = worklist.pop_back_val();
+      auto driver = ignoreHWDrivers ? RefDriver::get(val) : Driver::get(val);
```

Comment:
This is interesting, compared to what `CheckCombLoops` does, because here the drivers are needed only for the ports of the module, hence traversing the IR for each port even with the `getFieldRefFromValue` would potentially be non-overlapping traversal of the IR, which means a single IR traversal. Not 100% sure, but wondering if there is any difference between caching the result of `getFieldRefFromValue` , can it result in multiple traversals of the IR in some worst case scenario.

But `CheckCombLoops` first traverses the entire IR once to explicitly construct the "driver graph", (which I incorrectly named reaching def graph in the pass),  basic algorithm is, 
for each op in the module, update the "drivers graph" with the an edge from all the source(s) of the op to the destination of the op.

Reply:
Totally re:this chasing same indexing operations and caching likely being useful.  I'm unsure how useful caching `getFieldRefFromValue` would be since that would only help if the entire chain is re-walked (chain driven to multiple distinct destinations).  In my alternative implementation that's more powerful than this PR, it does more like what your CheckCombLoops does (as you describe), and also builds up a FieldRef-per-value cache as it walks (which I think CCL does too) (I think CCL does similar/same).

Creating the FieldRef's in the cache from a single `getFieldRefFromValue` walk would require tracking the chain as you go so you could get the final source Value and the fieldID for each -- so unsure it lends itself to caching in that way, offhand, or how much of a win that is? Maybe!  But totally it feels like this is possibly a bit more redundant here than is ideal :+1:.

You nailed it, this is a trade-off re:simplicity re:just chasing users but possibly re-walking same indexing operations VS walking everything once-ish but likely visting many operations that are irrelevant (and must analyze them all because you don't know a wire, for example, is going to be relevant until you're done).

Reply:
(easy enough to find out on a few designs, might as well! :+1:)

Reply:
Oops, was thinking of other version re:easy to test.  Here it's a bit harder because the driver subclasses are calling it (e.g., plumb in a cache or something).   Good points, leaving as-is for right now as it's very very unlikely to matter for probes and this isn't enabled for non-probes presently.

Reply:
Built a caching getFieldRefFromValue helper that I'll polish and submit after this, thanks for the suggestion!

Doesn't seem to help a lot in practice but there are sometimes fair amount of cache hits (even just densemap before `getFieldRefFromValue`, but also if cache intermediate indexing ops too) and it would help the worst-case performance as you say.